contracts/epochs/FlowDKG.cdc

```diff
@@ -0,0 +1,314 @@
+                    break
+                }
+
+                // update the index counter
+                finalSubmissionIndex = finalSubmissionIndex + 1
+            }
+
+            FlowDKG.finalSubmissionByNodeID[self.nodeID] = submission
+        }
+    }
+
+    // The Admin resource provides the ability to begin and end voting for an epoch
+    pub resource Admin {
+
+        /// Creates a new Participant resource for a consensus node
+        pub fun createParticipant(nodeID: String): @Participant {
+            let participant <-create Participant(nodeID: nodeID)
+            FlowDKG.nodeClaimed[nodeID] = true
+            return <-participant
+        }
+
+        /// Resets all the fields for tracking the current DKG process
+        /// and sets the given node IDs as registered
+        pub fun startDKG(nodeIDs: [String]) {
+            pre {
+                FlowDKG.dkgEnabled == false: "Cannot start the DKG when it is already running"
+            }
+            FlowDKG.dkgEnabled = true
+
+            FlowDKG.finalSubmissionByNodeID = {}
+            for id in nodeIDs {
+                FlowDKG.finalSubmissionByNodeID[id] = []
+            }
+
+            FlowDKG.whiteboardMessages = []
+
+            FlowDKG.uniqueFinalSubmissions = []
+
+            FlowDKG.uniqueFinalSubmissionCount = {}
+
+            emit StartDKG()
+        }
+
+        pub fun endDKG() {
+            pre { 
+                FlowDKG.dkgEnabled == true: "Cannot end the DKG when it is already disabled"
+                FlowDKG.dkgCompleted() != nil: "Cannot end the DKG until enough final arrays have been submitted"
+            }
+
+            FlowDKG.dkgEnabled = false
+
+            emit EndDKG(finalSubmission: FlowDKG.dkgCompleted())
+        }
+
+        /// Ends the DKG without checking if it is completed
+        /// Should only be used if the protocol halts and needs to be reset
+        pub fun forceEndDKG() {
+            FlowDKG.dkgEnabled = false
+
+            emit EndDKG(finalSubmission: FlowDKG.dkgCompleted())
+        }
+    }
+
+    /// Checks if two final submissions are equal by comparine each element
+    /// Each element has to be exactly the same
+    access(account) fun submissionsEqual(_ existingSubmission: [String], _ submission: [String]): Bool {
+
+        // If the submission length is different than the one being compared to, it is not equal
+        if submission.length != existingSubmission.length {
+            return false
+        }
+
+        var index = 0
+
+        // Check each key in the submiission to make sure that it matches
+        // the existing one
+        for key in submission {
+
+            // if a key is different, stop checking this submission
+            // and return false
+            if key != existingSubmission[index] {
+                return false
+            }
+
+            index = index + 1
```

Comment:
Would be nice to get the index when iterating in the for-loop, e.g `for index, key in submission`. Feel free to open a feature request

Reply:
Created! https://github.com/onflow/cadence/issues/781

Reply:
Awesome, thank you! 👍